Python/Product/Debugger/Debugger/PythonProcess.cs

```diff
@@ -1131,201 +907,176 @@ public PythonProcess(PythonLanguageVersion languageVersion, string exe, string a
-                    _stream.WriteInt32(unboundBreakpoint.Id);
-                    if (unboundBreakpoint.IsDjangoBreakpoint) {
-                        _stream.WriteString(unboundBreakpoint.Filename);
-                    }
-                }
+        internal async Task DisableBreakpointAsync(PythonBreakpoint unboundBreakpoint, CancellationToken ct) {
+            if (HasExited) {
+                return;
             }
-        }
 
-        internal void ConnectRepl(int portNum) {
-            DebugWriteCommand("Connect Repl");
-            lock (_streamLock) {
-                _stream.Write(ConnectReplCommandBytes);
-                _stream.WriteInt32(portNum);
-            }
+            await SendDebugRequestAsync(new LDP.RemoveBreakpointRequest() {
+                language = unboundBreakpoint.IsDjangoBreakpoint ? LDP.LanguageKind.Django : LDP.LanguageKind.Python,
+                breakpointId = unboundBreakpoint.Id,
+                breakpointFileName = unboundBreakpoint.Filename,
+                breakpointLineNo = unboundBreakpoint.LineNo,
+            }, ct);
         }
 
-        internal void DisconnectRepl() {
-            DebugWriteCommand("Disconnect Repl");
-            lock (_streamLock) {
-                if (_stream == null) {
-                    return;
-                }
+        internal async Task ConnectReplAsync(int portNum, CancellationToken ct = default(CancellationToken)) {
+            await SendDebugRequestAsync(new LDP.ConnectReplRequest() {
+                portNum = portNum,
+            }, ct);
+        }
 
-                try {
-                    _stream.Write(DisconnectReplCommandBytes);
-                } catch (IOException) {
-                } catch (ObjectDisposedException) {
-                    // If the process has terminated, we expect an exception
-                }
+        internal async Task DisconnectReplAsync(CancellationToken ct = default(CancellationToken)) {
+            try {
+                await SendDebugRequestAsync(new LDP.DisconnectReplRequest() {
+                }, ct);
+            } catch (IOException) {
+            } catch (ObjectDisposedException) {
+                // If the process has terminated, we expect an exception
             }
         }
 
-        internal bool SetLineNumber(PythonStackFrame pythonStackFrame, int lineNo) {
+        internal async Task<bool> SetLineNumberAsync(PythonStackFrame pythonStackFrame, int lineNo, CancellationToken ct) {
             if (_stoppedForException) {
                 return false;
             }
 
-            DebugWriteCommand("Set Line Number");
-            lock (_streamLock) {
-                _setLineResult = false;
-                _stream.Write(SetLineNumberCommand);
-                _stream.WriteInt64(pythonStackFrame.Thread.Id);
-                _stream.WriteInt32(pythonStackFrame.FrameId);
-                _stream.WriteInt32(lineNo);
-            }
+            var timeoutSource = new CancellationTokenSource(2000);
```

Comment:
I changed them all to use CancellationTokens. I've tested the 4 places where we use it in the product by forcing it to 0 ms. Error messages may not be ideal, but it doesn't break.

AD7Property.SetValueAsString: The value could not be set.  Please check your entry.

PythonProcess.SetLineNumberAsync: Unable to set the next statement. Operation not supported. Unknown error: 0x80004005.

AD7BoundBreakpoint.GetHitCount: No error from debugger, we let exception propagate out

AD7Property.EnumChildren: No error from debugger, we catch and return S_FALSE


---
In reply to: [107986767](https://github.com/Microsoft/PTVS/pull/2323#discussion_r107986767) [](ancestors = 107986767)